pkg/cmd/run/watch/watch.go

```diff
@@ -0,0 +1,228 @@
+
+	"github.com/cli/cli/api"
+	"github.com/cli/cli/internal/ghrepo"
+	"github.com/cli/cli/pkg/cmd/run/shared"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/cli/cli/utils"
+	"github.com/spf13/cobra"
+)
+
+const defaultInterval int = 3
+
+type WatchOptions struct {
+	IO         *iostreams.IOStreams
+	HttpClient func() (*http.Client, error)
+	BaseRepo   func() (ghrepo.Interface, error)
+
+	RunID      string
+	Interval   int
+	ExitStatus bool
+
+	Prompt bool
+
+	Now func() time.Time
+}
+
+func NewCmdWatch(f *cmdutil.Factory, runF func(*WatchOptions) error) *cobra.Command {
+	opts := &WatchOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+		Now:        time.Now,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "watch <run-selector>",
+		Short: "Runs until a run completes, showing its progress",
+		Annotations: map[string]string{
+			"IsActions": "true",
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// support `-R, --repo` override
+			opts.BaseRepo = f.BaseRepo
+
+			if len(args) > 0 {
+				opts.RunID = args[0]
+			} else if !opts.IO.CanPrompt() {
+				return &cmdutil.FlagError{Err: errors.New("run ID required when not running interactively")}
+			} else {
+				opts.Prompt = true
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return watchRun(opts)
+		},
+	}
+	cmd.Flags().BoolVar(&opts.ExitStatus, "exit-status", false, "Exit with non-zero status if run fails")
+	cmd.Flags().IntVarP(&opts.Interval, "interval", "i", defaultInterval, "Refresh interval in seconds")
+
+	return cmd
+}
+
+func watchRun(opts *WatchOptions) error {
+	c, err := opts.HttpClient()
+	if err != nil {
+		return fmt.Errorf("failed to create http client: %w", err)
+	}
+	client := api.NewClientFromHTTP(c)
+
+	repo, err := opts.BaseRepo()
+	if err != nil {
+		return fmt.Errorf("failed to determine base repo: %w", err)
+	}
+
+	runID := opts.RunID
+	var run *shared.Run
+
+	if opts.Prompt {
+		cs := opts.IO.ColorScheme()
+		runs, err := shared.GetRunsWithFilter(client, repo, 10, func(run shared.Run) bool {
+			return run.Status != shared.Completed
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get runs: %w", err)
+		}
+		if len(runs) == 0 {
+			return fmt.Errorf("found no in progress runs to watch")
+		}
+		runID, err = shared.PromptForRun(cs, runs)
+		if err != nil {
+			return err
+		}
+		// TODO silly stopgap until dust settles and PromptForRun can just return a run
+		for _, r := range runs {
+			if fmt.Sprintf("%d", r.ID) == runID {
+				run = &r
+				break
+			}
+		}
+	} else {
+		run, err = shared.GetRun(client, repo, runID)
+		if err != nil {
+			return fmt.Errorf("failed to get run: %w", err)
+		}
+	}
+
+	if run.Status == shared.Completed {
+		return nil
+	}
+
+	prNumber := ""
+	number, err := shared.PullRequestForRun(client, repo, *run)
+	if err == nil {
+		prNumber = fmt.Sprintf(" #%d", number)
+	}
+
+	if runtime.GOOS == "windows" {
+		opts.IO.EnableVirtualTerminalProcessing()
```

Comment:
I saw that we have conditional building for `console.go`. Why did you go that route instead of checking the OS inside of `opts.IO.EnableVirtualTerminalProcessing` and doing the correct thing? 

Additionally with the conditional building, is this check even needed? It look like a noop on non-windows OS.

Reply:
The conditional building is required so we can import the `windows` module; the code will not compile otherwise.

You're correct that the GOOS check is not needed for the call to `EnableVirtualTerminalProcessing` though.